fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteSumByAssociativeRule.java

```diff
@@ -0,0 +1,340 @@
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.starrocks.sql.optimizer.rule.transformation;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.starrocks.catalog.AggregateFunction;
+import com.starrocks.catalog.Function;
+import com.starrocks.catalog.FunctionSet;
+import com.starrocks.catalog.PrimitiveType;
+import com.starrocks.catalog.ScalarType;
+import com.starrocks.catalog.Type;
+import com.starrocks.server.GlobalStateMgr;
+import com.starrocks.sql.optimizer.OptExpression;
+import com.starrocks.sql.optimizer.OptimizerContext;
+import com.starrocks.sql.optimizer.base.ColumnRefFactory;
+import com.starrocks.sql.optimizer.base.ColumnRefSet;
+import com.starrocks.sql.optimizer.base.LogicalProperty;
+import com.starrocks.sql.optimizer.operator.OperatorType;
+import com.starrocks.sql.optimizer.operator.logical.LogicalAggregationOperator;
+import com.starrocks.sql.optimizer.operator.logical.LogicalProjectOperator;
+import com.starrocks.sql.optimizer.operator.pattern.Pattern;
+import com.starrocks.sql.optimizer.operator.scalar.CallOperator;
+import com.starrocks.sql.optimizer.operator.scalar.CastOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ColumnRefOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ConstantOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ScalarOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ScalarOperatorVisitor;
+import com.starrocks.sql.optimizer.rule.RuleType;
+
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.util.List;
+import java.util.Map;
+
+// For aggregate functions like `sum(expr + const)`, we can rewrite it to `sum(expr) + const * count()`
+// to reduce the overhead of a project calculation.
+public class RewriteSumByAssociativeRule extends TransformationRule {
+    public RewriteSumByAssociativeRule() {
+        super(RuleType.TF_REWRITE_SUM_BY_ASSOCIATIVE_RULE,
+                Pattern.create(OperatorType.LOGICAL_PROJECT)
+                        .addChildren(Pattern.create(OperatorType.LOGICAL_AGGR)
+                                .addChildren(Pattern.create(OperatorType.LOGICAL_PROJECT))));
+    }
+
+    @Override
+    public boolean check(final OptExpression input, OptimizerContext context) {
+        return !context.getSessionVariable().isDisableRewriteSumByAssociativeRule();
+    }
+
+    @Override
+    public List<OptExpression> transform(OptExpression input, OptimizerContext context) {
+        // let's use sum(id_int +1) as an example, suppose id_int is INT
```

Comment:
support other arithmetic expr? like: sum(id_int - 1), sum(id_int * 2) ?

Reply:
> support other arithmetic expr? like: sum(id_int - 1), sum(id_int * 2) ?

currently only support sum(expr + const) and sum(expr -const), I think we can use this idea to rewrite sum(expr * const), I will do it later